### PR TITLE
[Core] Make periodical runner thread safe.

### PR DESCRIPTION
**Pull request by rynewang**
Make periodical runner thread safe by using shared_ptr to manage a PeriodicalRunner's lifetime, and use weak_ptr for each task run to detect if the runner's alive.

Precedes #49047 .

### CODE REVIEW

File: src/ray/common/asio/periodical_runner.cc [hunk 1]
```diff
@@ -38,103 +33,106 @@ void PeriodicalRunner::Clear() {
...
+  timer->async_wait(
+      [weak_self = weak_from_this(),
+       fn = std::move(fn),
+       period,
+       timer = std::move(timer),
+       stats_handle = std::move(stats_handle),
+       name = std::move(name)](const boost::system::error_code &error) mutable {
+        if (auto self = weak_self.lock(); self) {
+          self->io_service_.stats().RecordExecution(
+              [weak_self,
```
Comment (jjyao) [inline src/ray/common/asio/periodical_runner.cc:115]: we can just capture `self` instead of `weak_self` since `RecordExecution` runs the lambda immediately.

File: src/ray/gcs/redis_context.cc [hunk 2]
```diff
@@ -673,7 +673,9 @@ void RedisContext::RunArgvAsync(std::vector<std::string> args,
                                                  std::move(redis_callback),
                                                  redis_async_context_.get(),
                                                  std::move(args));
-  request_context->Run();
+  // If we are already on the io_service thread, we can run the request immediately.
+  io_service_.dispatch([request_context]() { request_context->Run(); },
+                       "RedisContext::RunArgvAsync");
```
Comment (jjyao) [inline src/ray/gcs/redis_context.cc:678]: Unrelated change

File: src/ray/gcs/redis_context.h [hunk 3]
```diff
@@ -134,9 +135,15 @@ struct RedisRequestContext {
   std::vector<size_t> argc_;
 };
 
+// RunArgvAsync is thread-safe, can be accessed from multiple threads. The work is
+// dispatched to the io_service thread, and the callback is dispatched back to the
+// argument io_context.
+//
+// RunArgvSync is not thread-safe, caller must ensure there's no concurrent call to
+// RunArgvSync.
```
Comment (jjyao) [inline src/ray/gcs/redis_context.h:143]: Unrelated changes

File: src/ray/common/asio/periodical_runner.cc (hunk 1 above)
Comment (rynewang) [reply src/ray/common/asio/periodical_runner.cc:115]: done

File: src/ray/gcs/redis_context.h (hunk 3 above)
Comment (rynewang) [reply src/ray/gcs/redis_context.h]: done

File: src/ray/gcs/redis_context.cc (hunk 2 above)
Comment (rynewang) [reply src/ray/gcs/redis_context.cc:678]: done

File: src/ray/common/asio/periodical_runner.cc [hunk 4]
```diff
@@ -38,103 +33,106 @@ void PeriodicalRunner::Clear() {
...
+       fn = std::move(fn),
+       period,
+       timer = std::move(timer),
+       stats_handle = std::move(stats_handle),
+       name = std::move(name)](const boost::system::error_code &error) mutable {
+        if (auto self = weak_self.lock(); self) {
+          // Moves `self` to the lambda because RecordExecution runs the lambda
+          // immediately so there's no chance for `self` to go out of scope.
+          self->io_service_.stats().RecordExecution(
+              [self = std::move(self),
```
Comment (jjyao) [inline src/ray/common/asio/periodical_runner.cc:117]: This move is dangerous if line 117 runs before 116?

Comment (rynewang) [reply src/ray/common/asio/periodical_runner.cc]: you are right, changed to capture raw ptr

File: src/ray/common/asio/periodical_runner.cc [hunk 5]
```diff
@@ -38,103 +33,106 @@ void PeriodicalRunner::Clear() {
...
+       fn = std::move(fn),
+       period,
+       timer = std::move(timer),
+       stats_handle = std::move(stats_handle),
+       name = std::move(name)](const boost::system::error_code &error) mutable {
+        if (auto self = weak_self.lock(); self) {
+          // Captures `self_ptr` as a raw pointer because RecordExecution runs the lambda
+          // immediately so there's no chance for `self` to go out of scope.
+          self->io_service_.stats().RecordExecution(
+              [self_ptr = self.get(),
```
Comment (jjyao) [inline src/ray/common/asio/periodical_runner.cc:117]: I think it's safer to just do [self,

**Review by jjyao — APPROVED** [review]

File: src/ray/common/asio/periodical_runner.cc (hunk 5 above)
Comment (rynewang) [reply src/ray/common/asio/periodical_runner.cc:117]: well, ok
